DBZ-4409: Enforce consistent JSON serialization of vgtid

shichao-an opened the pull request:
Ensure `Vgtid.toString` method always produces consistent JSON string of the vgtid by making ShardGtids a list (respecting the list order from raw vgtid) and ensure key sequence in each JSON object representing a ShardGtid.

This fixes https://issues.redhat.com/browse/DBZ-4409

sonne5 commented:
Good find on the `ShardGtidSerializer` fix, it makes sense to me. Basically to make sure `gson.toJson` returns same result every time.

I wonder if we could get away without the change on sorted `ShardGtid`. The vgtid is extract [once per transaction](https://github.com/debezium/debezium-connector-vitess/blob/8fcef955ab32877dd34d23a5ceace03e873ae387/src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java#L90). Since json preserve the order of array, for every txn we should be able to share the same string representation. The order might change on different vstream responses for the same txn, but that seems fine for transaction metadata purpose.

Also thinking a bit more on using `vgtids` as the key for transactions, a concern I have is this would mean the txn key can be arbitrarily long since vgtid contains all the shards VStream is streaming from. I wonder if it makes sense to at least update the documentation to say:
> If we set `provide.transaction.metadata`, it is recommended to have one connector per shard set up and also set `stop_on_reshard` so that a connector won't consume multiple shards after shard split.

shichao-an commented on src/main/java/io/debezium/connector/vitess/Vgtid.java:
```diff
@@ -94,7 +106,7 @@ public int hashCode() {
     }
 
-    public static class ShardGtid {
+    public static class ShardGtid implements Comparable<ShardGtid> {
```
JsonPropertyOrder is Jackson I think?

gunnarmorling replied:
Yes, but it's a great point: we should use Jackson really, instead of Gson. Tbh. I hadn't noticed before that Gson would be used here at all, and it's the only connector which does use it. So for the sake of consistency, we should use Jackson indeed, also removing the Gson entry from the Debezium BOM then. Luckily, the Gson usage here is very contained, so this should be a quick change.

shichao-an replied:
Done.

gunnarmorling replied:
Perfect, thx!

shichao-an commented:
> I wonder if we could get away without the change on sorted `ShardGtid`. The vgtid is extract [once per transaction](https://github.com/debezium/debezium-connector-vitess/blob/8fcef955ab32877dd34d23a5ceace03e873ae387/src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java#L90). Since json preserve the order of array, for every txn we should be able to share the same string representation. The order might change on different vstream responses for the same txn, but that seems fine for transaction metadata purpose.

In that case, the problem is with Set for shardGtids here. We shouldn't use Set but a List if using raw vgtid as source of true.

> Also thinking a bit more on using `vgtids` as the key for transactions, a concern I have is this would mean the txn key can be arbitrarily long since vgtid contains all the shards VStream is streaming from. I wonder if it makes sense to at least update the documentation to say:
> 
> > If we set `provide.transaction.metadata`, it is recommended to have one connector per shard set up and also set `stop_on_reshard` so that a connector won't consume multiple shards after shard split.

That's a total different concern, but is valid. It's up to Vitess's semantics on vgtid. We can document recommended usage for sure.

shichao-an commented:
Updated shardGtids into a list @sonne5 
cc @jeffchao @keweishang

jeffchao commented:
LGTM but maybe in another PR this test could use a minor refactor to make the `shouldEquals` case more clear by separating out assertions into different cases; maybe use beforeEach as well.

shichao-an commented:
> LGTM but maybe in another PR this test could use a minor refactor to make the `shouldEquals` case more clear by separating out assertions into different cases; maybe use beforeEach as well.

I rename and refactored the test. I'm testing if equality holds if vgtid is created from different sources.

gunnarmorling reviewed (commented):
Looks good in general to me, some minor comments inline

We are planning to push out 1.8.0.CR1 tomorrow, so in case you would get that one wrapped up by then, we could include it in that release. The 1.8.0.Final release is planned for next week, so we could also add it to that one alternatively (we aim to minimize the changes between CR and Final, but there's some leeway for minor changes).

shichao-an commented on src/main/java/io/debezium/connector/vitess/Vgtid.java:
```diff
@@ -5,23 +5,32 @@
+    public static final String SHARD_KEY = "shard";
+    public static final String GTID_KEY = "gtid";
+
+    private static final Gson gson = new GsonBuilder().registerTypeAdapter(ShardGtid.class, new ShardGtidSerializer()).create();
```
Sounds good.

gunnarmorling replied:
Thx!

shichao-an commented:
> Looks good in general to me, some minor comments inline
> 
> We are planning to push out 1.8.0.CR1 tomorrow, so in case you would get that one wrapped up by then, we could include it in that release. The 1.8.0.Final release is planned for next week, so we could also add it to that one alternatively (we aim to minimize the changes between CR and Final, but there's some leeway for minor changes).

@gunnarmorling Thanks for reviewing and the info on release schedule. Addressed your comments.

gunnarmorling commented:
Btw. just to avoid any confusion due to time zones, above I meant to say the CR1 release is planned for "tomorrow, Thu 9th".

sonne5 commented on src/main/java/io/debezium/connector/vitess/Vgtid.java:
```diff
@@ -5,23 +5,30 @@
 public class Vgtid {
     public static final String CURRENT_GTID = "current";
-    private static final Gson gson = new Gson();
+    public static final String KEYSPACE_KEY = "keyspace";
+    public static final String SHARD_KEY = "shard";
+    public static final String GTID_KEY = "gtid";
```
```suggestion
    private static final String KEYSPACE_KEY = "keyspace";
    private static final String SHARD_KEY = "shard";
    private static final String GTID_KEY = "gtid";
```

shichao-an replied:
I made them public because if users want to deserialize Vgtid they can have access to the keys as well.

sonne5 reviewed (approved):
LGTM
